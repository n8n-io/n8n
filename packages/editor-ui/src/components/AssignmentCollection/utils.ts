import { isObject } from 'lodash-es';
import type { AssignmentValue, IDataObject } from 'n8n-workflow';
import { resolveParameter } from '@/composables/useWorkflowHelpers';
import { v4 as uuid } from 'uuid';

export function nameFromExpression(expression: string): string {
	return expression
		.replace(/^{{\s*|\s*}}$/g, '')
		.replace('$json.', '')
		.replace(/^\$\(.*\)(\.item\.json)?\.(.*)/, '$2');
}

export function inferAssignmentType(value: unknown): string {
	if (typeof value === 'boolean') return 'boolean';
	if (typeof value === 'number') return 'number';
	if (typeof value === 'string') return 'string';
	if (Array.isArray(value)) return 'array';
	if (isObject(value)) return 'object';
	return 'string';
}

export function typeFromExpression(expression: string): string {
	try {
		const resolved = resolveParameter(`=${expression}`);
		return inferAssignmentType(resolved);
	} catch (error) {
		return 'string';
	}
}

export function inputDataToAssignments(input: IDataObject): AssignmentValue[] {
	const assignments: AssignmentValue[] = [];

	function processValue(value: IDataObject, path: Array<string | number> = []) {
		if (Array.isArray(value)) {
			value.forEach((element, index) => {
				processValue(element, [...path, index]);
			});
		} else if (isObject(value)) {
			for (const [key, objectValue] of Object.entries(value)) {
				processValue(objectValue as IDataObject, [...path, key]);
			}
		} else {
			const stringPath = path.reduce((fullPath: string, part) => {
				if (typeof part === 'number') {
					return `${fullPath}[${part}]`;
				}
				return `${fullPath}.${part}`;
			}, '$json');

			const expression = `={{ ${stringPath} }}`;
			assignments.push({
				id: uuid(),
				name: stringPath.replace('$json.', ''),
				value: expression,
				type: inferAssignmentType(value),
			});
		}
	}

	processValue(input);

	return assignments;
}
