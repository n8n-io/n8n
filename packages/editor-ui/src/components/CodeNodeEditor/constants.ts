import { STICKY_NODE_TYPE } from '@/constants';
import type { Diagnostic } from '@codemirror/lint';

export const NODE_TYPES_EXCLUDED_FROM_AUTOCOMPLETION = [STICKY_NODE_TYPE];

export const AUTOCOMPLETABLE_BUILT_IN_MODULES = [
	'console',
	'constants',
	'crypto',
	'dns',
	'dns/promises',
	'fs',
	'fs/promises',
	'http',
	'http2',
	'https',
	'inspector',
	'module',
	'os',
	'path',
	'process',
	'readline',
	'url',
	'util',
	'zlib',
];

export const DEFAULT_LINTER_SEVERITY: Diagnostic['severity'] = 'error';

export const DEFAULT_LINTER_DELAY_IN_MS = 300;

/**
 * Length of the start of the script wrapper, used as offset for the linter to find a location in source text.
 */
export const OFFSET_FOR_SCRIPT_WRAPPER = 'module.exports = async function() {'.length;

export const ALL_ITEMS_PLACEHOLDER = `
// Loop over input items and add a new field
// called 'myNewField' to the JSON of each one
for (const item of $input.all()) {
  item.json.myNewField = 1;
}

return $input.all();
`.trim();

export const EACH_ITEM_PLACEHOLDER = `
// Add a new field called 'myNewField' to the
// JSON of the item
$input.item.json.myNewField = 1;

return $input.item;
`.trim();

export const CODE_LANGUAGES = ['javaScript', 'json'] as const;
export const CODE_MODES = ['runOnceForAllItems', 'runOnceForEachItem'] as const;
