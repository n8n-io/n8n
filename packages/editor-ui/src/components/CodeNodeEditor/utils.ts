import * as esprima from 'esprima-next';
import type { Completion } from '@codemirror/autocomplete';
import type { Node } from 'estree';
import type { RangeNode } from './types';

export function walk<T extends RangeNode>(
	node: Node | esprima.Program,
	test: (node: Node) => boolean,
	found: Node[] = [],
) {
	// @ts-ignore
	if (test(node)) found.push(node);

	for (const key in node) {
		if (!(key in node)) continue;

		// @ts-ignore
		const child = node[key];

		if (child === null || typeof child !== 'object') continue;

		if (Array.isArray(child)) {
			child.forEach((node) => walk(node, test, found));
		} else {
			walk(child, test, found);
		}
	}

	return found as T[];
}

export const escape = (str: string) =>
	str
		.replace('$', '\\$')
		.replace('(', '\\(')
		.replace(')', '\\)')
		.replace('[', '\\[')
		.replace(']', '\\]');

export const toVariableOption = (label: string) => ({ label, type: 'variable' });

export const addVarType = (option: Completion) => ({ ...option, type: 'variable' });
