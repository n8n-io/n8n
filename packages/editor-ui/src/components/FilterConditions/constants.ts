/* eslint-disable @typescript-eslint/naming-convention */
import type { FilterConditionValue, FilterOptionsValue } from 'n8n-workflow';
import type { FilterOperator, FilterOperatorGroup } from './types';

export const DEFAULT_MAX_CONDITIONS = 10;

export const DEFAULT_FILTER_OPTIONS: FilterOptionsValue = {
	caseSensitive: true,
	leftValue: '',
	typeValidation: 'strict',
};

export const OPERATORS_BY_ID = {
	'string:exists': {
		type: 'string',
		operation: 'exists',
		name: 'filter.operator.exists',
		singleValue: true,
	},
	'string:notExists': {
		type: 'string',
		operation: 'notExists',
		name: 'filter.operator.notExists',
		singleValue: true,
	},
	'string:empty': {
		type: 'string',
		operation: 'empty',
		name: 'filter.operator.empty',
		singleValue: true,
	},
	'string:notEmpty': {
		type: 'string',
		operation: 'notEmpty',
		name: 'filter.operator.notEmpty',
		singleValue: true,
	},
	'string:equals': { type: 'string', operation: 'equals', name: 'filter.operator.equals' },
	'string:notEquals': { type: 'string', operation: 'notEquals', name: 'filter.operator.notEquals' },
	'string:contains': { type: 'string', operation: 'contains', name: 'filter.operator.contains' },
	'string:notContains': {
		type: 'string',
		operation: 'notContains',
		name: 'filter.operator.notContains',
	},
	'string:startsWith': {
		type: 'string',
		operation: 'startsWith',
		name: 'filter.operator.startsWith',
	},
	'string:notStartsWith': {
		type: 'string',
		operation: 'notStartsWith',
		name: 'filter.operator.notStartsWith',
	},
	'string:endsWith': { type: 'string', operation: 'endsWith', name: 'filter.operator.endsWith' },
	'string:notEndsWith': {
		type: 'string',
		operation: 'notEndsWith',
		name: 'filter.operator.notEndsWith',
	},
	'string:regex': { type: 'string', operation: 'regex', name: 'filter.operator.regex' },
	'string:notRegex': { type: 'string', operation: 'notRegex', name: 'filter.operator.notRegex' },
	'number:exists': {
		type: 'number',
		operation: 'exists',
		name: 'filter.operator.exists',
		singleValue: true,
	},
	'number:notExists': {
		type: 'number',
		operation: 'notExists',
		name: 'filter.operator.notExists',
		singleValue: true,
	},
	'number:equals': { type: 'number', operation: 'equals', name: 'filter.operator.equals' },
	'number:notEquals': { type: 'number', operation: 'notEquals', name: 'filter.operator.notEquals' },
	'number:gt': { type: 'number', operation: 'gt', name: 'filter.operator.gt' },
	'number:lt': { type: 'number', operation: 'lt', name: 'filter.operator.lt' },
	'number:gte': { type: 'number', operation: 'gte', name: 'filter.operator.gte' },
	'number:lte': { type: 'number', operation: 'lte', name: 'filter.operator.lte' },
	'dateTime:exists': {
		type: 'dateTime',
		operation: 'exists',
		name: 'filter.operator.exists',
		singleValue: true,
	},
	'dateTime:notExists': {
		type: 'dateTime',
		operation: 'notExists',
		name: 'filter.operator.notExists',
		singleValue: true,
	},
	'dateTime:equals': { type: 'dateTime', operation: 'equals', name: 'filter.operator.equals' },
	'dateTime:notEquals': {
		type: 'dateTime',
		operation: 'notEquals',
		name: 'filter.operator.notEquals',
	},
	'dateTime:after': { type: 'dateTime', operation: 'after', name: 'filter.operator.after' },
	'dateTime:before': { type: 'dateTime', operation: 'before', name: 'filter.operator.before' },
	'dateTime:afterOrEquals': {
		type: 'dateTime',
		operation: 'afterOrEquals',
		name: 'filter.operator.afterOrEquals',
	},
	'dateTime:beforeOrEquals': {
		type: 'dateTime',
		operation: 'beforeOrEquals',
		name: 'filter.operator.beforeOrEquals',
	},
	'boolean:exists': {
		type: 'boolean',
		operation: 'exists',
		name: 'filter.operator.exists',
		singleValue: true,
	},
	'boolean:notExists': {
		type: 'boolean',
		operation: 'notExists',
		name: 'filter.operator.notExists',
		singleValue: true,
	},
	'boolean:true': {
		type: 'boolean',
		operation: 'true',
		name: 'filter.operator.true',
		singleValue: true,
	},
	'boolean:false': {
		type: 'boolean',
		operation: 'false',
		name: 'filter.operator.false',
		singleValue: true,
	},
	'boolean:equals': { type: 'boolean', operation: 'equals', name: 'filter.operator.equals' },
	'boolean:notEquals': {
		type: 'boolean',
		operation: 'notEquals',
		name: 'filter.operator.notEquals',
	},
	'array:exists': {
		type: 'array',
		operation: 'exists',
		name: 'filter.operator.exists',
		singleValue: true,
	},
	'array:notExists': {
		type: 'array',
		operation: 'notExists',
		name: 'filter.operator.notExists',
		singleValue: true,
	},
	'array:empty': {
		type: 'array',
		operation: 'empty',
		name: 'filter.operator.empty',
		singleValue: true,
	},
	'array:notEmpty': {
		type: 'array',
		operation: 'notEmpty',
		name: 'filter.operator.notEmpty',
		singleValue: true,
	},
	'array:contains': {
		type: 'array',
		operation: 'contains',
		name: 'filter.operator.contains',
		rightType: 'any',
	},
	'array:notContains': {
		type: 'array',
		operation: 'notContains',
		name: 'filter.operator.notContains',
		rightType: 'any',
	},
	'array:lengthEquals': {
		type: 'array',
		operation: 'lengthEquals',
		name: 'filter.operator.lengthEquals',
		rightType: 'number',
	},
	'array:lengthNotEquals': {
		type: 'array',
		operation: 'lengthNotEquals',
		name: 'filter.operator.lengthNotEquals',
		rightType: 'number',
	},
	'array:lengthGt': {
		type: 'array',
		operation: 'lengthGt',
		name: 'filter.operator.lengthGt',
		rightType: 'number',
	},
	'array:lengthLt': {
		type: 'array',
		operation: 'lengthLt',
		name: 'filter.operator.lengthLt',
		rightType: 'number',
	},
	'array:lengthGte': {
		type: 'array',
		operation: 'lengthGte',
		name: 'filter.operator.lengthGte',
		rightType: 'number',
	},
	'array:lengthLte': {
		type: 'array',
		operation: 'lengthLte',
		name: 'filter.operator.lengthLte',
		rightType: 'number',
	},
	'object:exists': {
		type: 'object',
		operation: 'exists',
		name: 'filter.operator.exists',
		singleValue: true,
	},
	'object:notExists': {
		type: 'object',
		operation: 'notExists',
		name: 'filter.operator.notExists',
		singleValue: true,
	},
	'object:empty': {
		type: 'object',
		operation: 'empty',
		name: 'filter.operator.empty',
		singleValue: true,
	},
	'object:notEmpty': {
		type: 'object',
		operation: 'notEmpty',
		name: 'filter.operator.notEmpty',
		singleValue: true,
	},
} as const satisfies Record<string, FilterOperator>;

export const OPERATORS = Object.values(OPERATORS_BY_ID);

export type FilterOperatorId = keyof typeof OPERATORS_BY_ID;

export const DEFAULT_OPERATOR_VALUE: FilterConditionValue['operator'] =
	OPERATORS_BY_ID['string:equals'];

export const OPERATOR_GROUPS: FilterOperatorGroup[] = [
	{
		id: 'string',
		name: 'type.string',
		icon: 'font',
		children: OPERATORS.filter((operator) => operator.type === 'string'),
	},
	{
		id: 'number',
		name: 'type.number',
		icon: 'hashtag',
		children: OPERATORS.filter((operator) => operator.type === 'number'),
	},
	{
		id: 'dateTime',
		name: 'type.dateTime',
		icon: 'calendar',
		children: OPERATORS.filter((operator) => operator.type === 'dateTime'),
	},
	{
		id: 'boolean',
		name: 'type.boolean',
		icon: 'check-square',
		children: OPERATORS.filter((operator) => operator.type === 'boolean'),
	},
	{
		id: 'array',
		name: 'type.array',
		icon: 'list',
		children: OPERATORS.filter((operator) => operator.type === 'array'),
	},
	{
		id: 'object',
		name: 'type.object',
		icon: 'cube',
		children: OPERATORS.filter((operator) => operator.type === 'object'),
	},
];
