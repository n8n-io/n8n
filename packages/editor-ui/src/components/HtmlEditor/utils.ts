import type { Range } from './types';

/**
 * Return the ranges of a full range that are _not_ within the taken ranges,
 * assuming sorted taken ranges. e.g. `[0, 10]` and `[[2, 3], [7, 8]]`
 * return `[[0, 1], [4, 6], [9, 10]]`
 */
export function nonTakenRanges(fullRange: Range, takenRanges: Range[]) {
	const found = [];

	const [fullStart, fullEnd] = fullRange;
	let i = fullStart;
	let curStart = fullStart;

	takenRanges = [...takenRanges];

	while (i < fullEnd) {
		if (takenRanges.length === 0) {
			found.push([curStart, fullEnd]);
			break;
		}

		const [takenStart, takenEnd] = takenRanges[0];

		if (i < takenStart) {
			i++;
			continue;
		}

		if (takenStart !== fullStart) {
			found.push([curStart, i - 1]);
		}

		i = takenEnd + 1;
		curStart = takenEnd + 1;
		takenRanges.shift();
	}

	return found;
}
