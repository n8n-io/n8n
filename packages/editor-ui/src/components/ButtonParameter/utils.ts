import type { Schema } from '@/Interface';
import { ApplicationError, type INodeExecutionData } from 'n8n-workflow';
import { useWorkflowsStore } from '@/stores/workflows.store';
import { useNDVStore } from '@/stores/ndv.store';
import { useDataSchema } from '@/composables/useDataSchema';
import { executionDataToJson } from '@/utils/nodeTypesUtils';
import { generateCodeForPrompt } from '../../api/ai';
import { useRootStore } from '../../stores/root.store';
import { type AskAiRequest } from '../../types/assistant.types';
import { useSettingsStore } from '../../stores/settings.store';
import { format } from 'prettier';
import jsParser from 'prettier/plugins/babel';
import * as estree from 'prettier/plugins/estree';

export function getParentNodes() {
	const activeNode = useNDVStore().activeNode;
	const { getCurrentWorkflow, getNodeByName } = useWorkflowsStore();
	const workflow = getCurrentWorkflow();

	if (!activeNode || !workflow) return [];

	return workflow
		.getParentNodesByDepth(activeNode?.name)
		.filter(({ name }, i, nodes) => {
			return name !== activeNode.name && nodes.findIndex((node) => node.name === name) === i;
		})
		.map((n) => getNodeByName(n.name))
		.filter((n) => n !== null);
}

export function getSchemas() {
	const parentNodes = getParentNodes();
	const parentNodesNames = parentNodes.map((node) => node?.name);
	const { getSchemaForExecutionData, getInputDataWithPinned } = useDataSchema();
	const parentNodesSchemas: Array<{ nodeName: string; schema: Schema }> = parentNodes
		.map((node) => {
			const inputData: INodeExecutionData[] = getInputDataWithPinned(node);

			return {
				nodeName: node?.name || '',
				schema: getSchemaForExecutionData(executionDataToJson(inputData), true),
			};
		})
		.filter((node) => node.schema?.value.length > 0);

	const inputSchema = parentNodesSchemas.shift();

	return {
		parentNodesNames,
		inputSchema,
		parentNodesSchemas,
	};
}

export async function generateCodeForAiTransform(prompt: string, path: string) {
	const schemas = getSchemas();

	const payload: AskAiRequest.RequestPayload = {
		question: prompt,
		context: {
			schema: schemas.parentNodesSchemas,
			inputSchema: schemas.inputSchema!,
			ndvPushRef: useNDVStore().pushRef,
			pushRef: useRootStore().pushRef,
		},
		forNode: 'transform',
	};

	let value;
	if (useSettingsStore().isAskAiEnabled) {
		const { restApiContext } = useRootStore();
		const { code } = await generateCodeForPrompt(restApiContext, payload);
		value = code;
	} else {
		throw new ApplicationError('AI code generation is not enabled');
	}

	if (value === undefined) return;

	const formattedCode = await format(String(value), {
		parser: 'babel',
		plugins: [jsParser, estree],
	});

	const updateInformation = {
		name: path,
		value: formattedCode,
	};

	return updateInformation;
}
