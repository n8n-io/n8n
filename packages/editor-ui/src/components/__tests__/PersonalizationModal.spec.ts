import { PiniaVuePlugin } from 'pinia';
import { createLocalVue, mount } from '@vue/test-utils';
import PersonalizationModal from '@/components/PersonalizationModal.vue';
import { createTestingPinia } from '@pinia/testing';
import { PERSONALIZATION_MODAL_KEY } from '@/constants';
import { retry } from '@/__tests__/utils';

describe('PersonalizationModal.vue', () => {
	const pinia = createTestingPinia({
		initialState: {
			ui: {
				modals: {
					[PERSONALIZATION_MODAL_KEY]: { open: true },
				},
			},
			settings: {
				settings: {
					templates: {
						host: '',
					},
				},
			},
		},
	});
	const localVue = createLocalVue();
	localVue.use(PiniaVuePlugin);

	it('should render correctly', async () => {
		const wrapper = mount(PersonalizationModal, {
			localVue,
			pinia,
		});

		await retry(() => expect(wrapper.find('.modal-content').exists()).toBe(true));

		expect(wrapper.findAll('.n8n-select').length).toEqual(5);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should display new option when role is "Devops", "Engineering", "IT", or "Sales and marketing"', async () => {
		const wrapper = mount(PersonalizationModal, {
			localVue,
			pinia,
		});

		await retry(() => expect(wrapper.find('.modal-content').exists()).toBe(true));

		for (const index of [3, 4, 5, 6]) {
			wrapper.find('.n8n-select[name="role"]').trigger('click');
			wrapper
				.find('.n8n-select[name="role"]')
				.findAll('.el-select-dropdown__item')
				.at(index)
				.trigger('click');

			await retry(() => {
				expect(wrapper.findAll('.n8n-select').length).toEqual(6);
				expect(wrapper.find('.n8n-select[name^="automationGoal"]').exists()).toBe(true);
			});
		}
	});
});
