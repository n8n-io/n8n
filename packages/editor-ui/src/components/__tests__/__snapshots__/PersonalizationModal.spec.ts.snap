// Vitest Snapshot v1

exports[`PersonalizationModal.vue > should render correctly 1`] = `
"<transition-stub name=\\"dialog-fade\\" data-test-id=\\"personalization-form\\" class=\\"dialog-wrapper\\" style=\\"z-index: 2001;\\">
  <div class=\\"el-dialog__wrapper\\">
    <div role=\\"dialog\\" aria-modal=\\"true\\" aria-label=\\"dialog\\" class=\\"el-dialog\\" style=\\"margin-top: 15vh; width: 460px;\\">
      <div class=\\"el-dialog__header\\">
        <div class=\\"centerTitle\\">
          <div>
            <h1 class=\\"n8n-heading size-xlarge regular\\">Customize n8n to you</h1>
          </div>
          <div class=\\"subtitle\\">
            <h3 class=\\"n8n-heading text-light size-small regular\\">These questions help us tailor n8n to you</h3>
          </div>
        </div>
        <!---->
      </div>
      <div class=\\"el-dialog__body\\">
        <div class=\\"modal-content\\">
          <div class=\\"container\\">
            <div>
              <div class=\\"grid\\">
                <div class=\\"\\">
                  <div class=\\"container\\" data-test-id=\\"companyType\\"><label for=\\"companyType\\" class=\\"inputLabel heading medium\\">
                      <div class=\\"title\\"><span class=\\"n8n-text size-medium bold\\"> What best describes your company? <!----></span></div>
                      <!---->
                      <!---->
                      <!---->
                    </label>
                    <div class=\\"\\">
                      <div class=\\"n8n-select container\\" name=\\"companyType\\">
                        <!---->
                        <div class=\\"el-select el-select--large\\">
                          <!---->
                          <div class=\\"el-input el-input--large el-input--suffix\\">
                            <!----><input type=\\"text\\" readonly=\\"readonly\\" autocomplete=\\"off\\" placeholder=\\"Select...\\" class=\\"el-input__inner\\">
                            <!----><span class=\\"el-input__suffix\\"><span class=\\"el-input__suffix-inner\\"><i class=\\"el-select__caret el-input__icon el-icon-arrow-up\\"></i><!----><!----><!----><!----><!----></span>
                            <!----></span>
                            <!---->
                            <!---->
                          </div>
                          <transition-stub name=\\"el-zoom-in-top\\">
                            <div class=\\"el-select-dropdown el-popper\\" style=\\"display: none;\\">
                              <div class=\\"el-scrollbar\\" style=\\"\\">
                                <div class=\\"el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default\\">
                                  <ul class=\\"el-scrollbar__view el-select-dropdown__list\\">
                                    <!---->
                                    <li class=\\"el-select-dropdown__item\\"><span>Software as a service</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Ecommerce</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Marketing agency / consultancy</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Systems integrator / Automation agency</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Education</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Other</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>I'm not using n8n for work</span></li>
                                  </ul>
                                </div>
                                <div class=\\"el-scrollbar__bar is-horizontal\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateX(0%); webkit-transform: translateX(0%);\\"></div>
                                </div>
                                <div class=\\"el-scrollbar__bar is-vertical\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateY(0%); webkit-transform: translateY(0%);\\"></div>
                                </div>
                              </div>
                              <!---->
                            </div>
                          </transition-stub>
                        </div>
                      </div>
                    </div>
                    <!---->
                  </div>
                </div>
                <div class=\\"mt-undefined\\">
                  <div class=\\"container\\" data-test-id=\\"role\\"><label for=\\"role\\" class=\\"inputLabel heading medium\\">
                      <div class=\\"title\\"><span class=\\"n8n-text size-medium bold\\"> Which role best describes you? <!----></span></div>
                      <!---->
                      <!---->
                      <!---->
                    </label>
                    <div class=\\"\\">
                      <div class=\\"n8n-select container\\" name=\\"role\\">
                        <!---->
                        <div class=\\"el-select el-select--large\\">
                          <!---->
                          <div class=\\"el-input el-input--large el-input--suffix\\">
                            <!----><input type=\\"text\\" readonly=\\"readonly\\" autocomplete=\\"off\\" placeholder=\\"Select...\\" class=\\"el-input__inner\\">
                            <!----><span class=\\"el-input__suffix\\"><span class=\\"el-input__suffix-inner\\"><i class=\\"el-select__caret el-input__icon el-icon-arrow-up\\"></i><!----><!----><!----><!----><!----></span>
                            <!----></span>
                            <!---->
                            <!---->
                          </div>
                          <transition-stub name=\\"el-zoom-in-top\\">
                            <div class=\\"el-select-dropdown el-popper\\" style=\\"display: none;\\">
                              <div class=\\"el-scrollbar\\" style=\\"\\">
                                <div class=\\"el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default\\">
                                  <ul class=\\"el-scrollbar__view el-select-dropdown__list\\">
                                    <!---->
                                    <li class=\\"el-select-dropdown__item\\"><span>Business Owner</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Customer support</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Data Science</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Devops</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>IT</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Engineering</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Sales and Marketing</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Security</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Other (please specify)</span></li>
                                  </ul>
                                </div>
                                <div class=\\"el-scrollbar__bar is-horizontal\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateX(0%); webkit-transform: translateX(0%);\\"></div>
                                </div>
                                <div class=\\"el-scrollbar__bar is-vertical\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateY(0%); webkit-transform: translateY(0%);\\"></div>
                                </div>
                              </div>
                              <!---->
                            </div>
                          </transition-stub>
                        </div>
                      </div>
                    </div>
                    <!---->
                  </div>
                </div>
                <div class=\\"mt-undefined\\">
                  <div class=\\"container\\" data-test-id=\\"automationBeneficiary\\"><label for=\\"automationBeneficiary\\" class=\\"inputLabel heading medium\\">
                      <div class=\\"title\\"><span class=\\"n8n-text size-medium bold\\"> Who will your automations mainly be for? <!----></span></div>
                      <!---->
                      <!---->
                      <!---->
                    </label>
                    <div class=\\"\\">
                      <div class=\\"n8n-select container\\" name=\\"automationBeneficiary\\">
                        <!---->
                        <div class=\\"el-select el-select--large\\">
                          <!---->
                          <div class=\\"el-input el-input--large el-input--suffix\\">
                            <!----><input type=\\"text\\" readonly=\\"readonly\\" autocomplete=\\"off\\" placeholder=\\"Select...\\" class=\\"el-input__inner\\">
                            <!----><span class=\\"el-input__suffix\\"><span class=\\"el-input__suffix-inner\\"><i class=\\"el-select__caret el-input__icon el-icon-arrow-up\\"></i><!----><!----><!----><!----><!----></span>
                            <!----></span>
                            <!---->
                            <!---->
                          </div>
                          <transition-stub name=\\"el-zoom-in-top\\">
                            <div class=\\"el-select-dropdown el-popper\\" style=\\"display: none;\\">
                              <div class=\\"el-scrollbar\\" style=\\"\\">
                                <div class=\\"el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default\\">
                                  <ul class=\\"el-scrollbar__view el-select-dropdown__list\\">
                                    <!---->
                                    <li class=\\"el-select-dropdown__item\\"><span>Myself</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>My team</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Other teams</span></li>
                                  </ul>
                                </div>
                                <div class=\\"el-scrollbar__bar is-horizontal\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateX(0%); webkit-transform: translateX(0%);\\"></div>
                                </div>
                                <div class=\\"el-scrollbar__bar is-vertical\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateY(0%); webkit-transform: translateY(0%);\\"></div>
                                </div>
                              </div>
                              <!---->
                            </div>
                          </transition-stub>
                        </div>
                      </div>
                    </div>
                    <!---->
                  </div>
                </div>
                <div class=\\"mt-undefined\\">
                  <div class=\\"container\\" data-test-id=\\"companySize\\"><label for=\\"companySize\\" class=\\"inputLabel heading medium\\">
                      <div class=\\"title\\"><span class=\\"n8n-text size-medium bold\\"> How big is your company? <!----></span></div>
                      <!---->
                      <!---->
                      <!---->
                    </label>
                    <div class=\\"\\">
                      <div class=\\"n8n-select container\\" name=\\"companySize\\">
                        <!---->
                        <div class=\\"el-select el-select--large\\">
                          <!---->
                          <div class=\\"el-input el-input--large el-input--suffix\\">
                            <!----><input type=\\"text\\" readonly=\\"readonly\\" autocomplete=\\"off\\" placeholder=\\"Select...\\" class=\\"el-input__inner\\">
                            <!----><span class=\\"el-input__suffix\\"><span class=\\"el-input__suffix-inner\\"><i class=\\"el-select__caret el-input__icon el-icon-arrow-up\\"></i><!----><!----><!----><!----><!----></span>
                            <!----></span>
                            <!---->
                            <!---->
                          </div>
                          <transition-stub name=\\"el-zoom-in-top\\">
                            <div class=\\"el-select-dropdown el-popper\\" style=\\"display: none;\\">
                              <div class=\\"el-scrollbar\\" style=\\"\\">
                                <div class=\\"el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default\\">
                                  <ul class=\\"el-scrollbar__view el-select-dropdown__list\\">
                                    <!---->
                                    <li class=\\"el-select-dropdown__item\\"><span>Less than 20 people</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>20-99 people</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>100-499 people</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>500-999 people</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>1000+ people</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>I'm not using n8n for work</span></li>
                                  </ul>
                                </div>
                                <div class=\\"el-scrollbar__bar is-horizontal\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateX(0%); webkit-transform: translateX(0%);\\"></div>
                                </div>
                                <div class=\\"el-scrollbar__bar is-vertical\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateY(0%); webkit-transform: translateY(0%);\\"></div>
                                </div>
                              </div>
                              <!---->
                            </div>
                          </transition-stub>
                        </div>
                      </div>
                    </div>
                    <!---->
                  </div>
                </div>
                <div class=\\"mt-undefined\\">
                  <div class=\\"container\\" data-test-id=\\"reportedSource\\"><label for=\\"reportedSource\\" class=\\"inputLabel heading medium\\">
                      <div class=\\"title\\"><span class=\\"n8n-text size-medium bold\\"> How did you hear about n8n? <!----></span></div>
                      <!---->
                      <!---->
                      <!---->
                    </label>
                    <div class=\\"\\">
                      <div class=\\"n8n-select container\\" name=\\"reportedSource\\">
                        <!---->
                        <div class=\\"el-select el-select--large\\">
                          <!---->
                          <div class=\\"el-input el-input--large el-input--suffix\\">
                            <!----><input type=\\"text\\" readonly=\\"readonly\\" autocomplete=\\"off\\" placeholder=\\"Select...\\" class=\\"el-input__inner\\">
                            <!----><span class=\\"el-input__suffix\\"><span class=\\"el-input__suffix-inner\\"><i class=\\"el-select__caret el-input__icon el-icon-arrow-up\\"></i><!----><!----><!----><!----><!----></span>
                            <!----></span>
                            <!---->
                            <!---->
                          </div>
                          <transition-stub name=\\"el-zoom-in-top\\">
                            <div class=\\"el-select-dropdown el-popper\\" style=\\"display: none;\\">
                              <div class=\\"el-scrollbar\\" style=\\"\\">
                                <div class=\\"el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default\\">
                                  <ul class=\\"el-scrollbar__view el-select-dropdown__list\\">
                                    <!---->
                                    <li class=\\"el-select-dropdown__item\\"><span>Google</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Twitter</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>LinkedIn</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>YouTube</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Friend / Word of mouth</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Podcast</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Event</span></li>
                                    <li class=\\"el-select-dropdown__item\\"><span>Other (please specify)</span></li>
                                  </ul>
                                </div>
                                <div class=\\"el-scrollbar__bar is-horizontal\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateX(0%); webkit-transform: translateX(0%);\\"></div>
                                </div>
                                <div class=\\"el-scrollbar__bar is-vertical\\">
                                  <div class=\\"el-scrollbar__thumb\\" style=\\"transform: translateY(0%); webkit-transform: translateY(0%);\\"></div>
                                </div>
                              </div>
                              <!---->
                            </div>
                          </transition-stub>
                        </div>
                      </div>
                    </div>
                    <!---->
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class=\\"footer\\">
          <div><button aria-disabled=\\"false\\" aria-busy=\\"false\\" aria-live=\\"polite\\" class=\\"button button primary medium float-right\\">
              <!----><span>Continue</span></button></div>
        </div>
      </div>
      <!---->
    </div>
  </div>
</transition-stub>"
`;
